### ISSUE TITLE
be-gateway quickstarter

### ISSUE THREAD

**Issue by gerardcl** [opendevstack/ods-project-quickstarters#427]
**Is your feature request related to a problem? Please describe.**
Since we provide different kind of quickstarters (usually following microservice architecture approach) which usually are dependant between each other, it would be nice to have available an API/Gateway quickstarter that can really handle huge loads and different requirements (e.g.: Nginx).
This is also related to https://github.com/opendevstack/ods-project-quickstarters/issues/424 and 
https://github.com/opendevstack/ods-core/issues/215

**Describe the solution you'd like**
Provide a be-gateway quickstarter with a set of ready-to-use Lua modules for different API/Gateway requirements. This would be using OpenResty.

**Describe alternatives you've considered**
To not provide it --> users will have to use another solution for such requirement

**Additional context**
To proceed by following official images provided --> https://github.com/openresty/docker-openresty

**Comment by rattermeyer** [comment]
We are currently considering Kong. We have several use cases where we need Kong. 
We are currently exploring different ways to include Kong.
@tjaeschke is currently looking into it.

**Comment by gerardcl** [comment]
Hi @rattermeyer and @tjaeschke !
Nice!
Have you considered [KrakenD](https://www.krakend.io/) ? I would strongly recommend using it! Check the features and benchmarks :) 

If you want I could come up with a PoC, what do you think?

Best,

**Comment by stitakis** [comment]
Hi @gerardcl and @rattermeyer, to shed some light, can you please describe what use cases you need or would like to address? Asking because I think it will simplify the discussion and put the focus in finding solution.

**Comment by metmajer** [comment]
like the approach @stitakis, but looking at a handful of use-cases probably isn’t enough to derive a standard solution to cover a solution space that‘s still young and growing. Hence, consider going with a named approach for now, as in `be-gateway-$techX` to cater different needs. Did I mention that I like solo.io‘s Gloo? 😁

**Comment by stitakis** [comment]
@metmajer With ODS 2.0 will be possible to provision your own quickstarters by pointing to your own git repo. This would make possible to anyone provision it own custom stuff. A great feature coming with ODS 2.0. (@michaelsauter pls correct me if I'm wrong) 
However, what kind of quickstarter ODS should provide is a different question. I see here the chance to list the use cases for gateway and come up with a solution that covers at least the majority of them. So that users can opt, to use standard gateway from ods or provision they own version. Does this make sense?

**Comment by michaelsauter** [comment]
@stitakis Correct, that is the plan and a long-awaited feature I would say. My take would be that ODS should probably provide one solution "officially"/"out-of-the-box", and right now I have no preference personally. On top of that, we/others could have more quickstarters with different tech that cater the same/similar use-cases. Therefore, totally agree with @metmajer that the used tech should be in the quickstarter name.

For 2.0, it would be great to unify the quickstarter names and ensure some kind of pattern - different issue though :)

**Comment by tjaeschke** [comment]
I think we should not focus on one "official" solution, but support multiple API gateway implementations if possible, as long as this doesn't focus on a special use-case. 

Kong has the benefit, that it is a RedHat certified software and the official image is OpenShift ready.

I agree with @metmajer and @michaelsauter that the tech should be reflected int the quickstarter name.

**Comment by michaelsauter** [comment]
With "official" I meant in the repo under ODS ... IMHO it would be good for users to have one "default" solution. Kong sounds like a good fit if it is close to OpenShift ...

**Comment by gerardcl** [comment]
Hi! 

Didn't know about the RedHat certified badge for Kong, cool! then I would vote for that as a default option. 

I fully agree on the naming convention if more than one tech/solution per use-case is possible. So, even for a Kong quickstarter a might go on calling it -> be-gateway-kong, or similar order.

For me API Gateway use-cases would be all expected features an API Gateway needs to have as of definition (see for example any feature list of any of the already listed in this conversation, they all offer similar features but maybe different workflows/implementations...but the requirements are all the same since API Gateway is a defined concept...that we do not need to re-invent).

**Comment by henrjk** [comment]
Hey, sorry for my ignorance. I don't know what the crisp definition of the API Gateway would be? 

Just reading https://docs.microsoft.com/en-us/azure/architecture/microservices/design/gateway there are several patterns suggested (Gateway routing, GW Aggregation,  GW Offloading) and a wide array of functionality that could be offloaded.

Would be great if one had a number of gateways to choose from and easily understand how these differ. 

In my mind one intent of this issue is to not include security related filtering as this can impede open source contributions. @gerardcl does the last statement makes sense to you?

**Comment by gerardcl** [comment]
hi @henrjk 

I fully see your point(s) and agree!

I think that the (somehow agreed) aforementioned in this conversation regarding naming convention would fit too.

**Comment by jonkristjan** [comment]
Hi everyone, 

Love the discussion. :) I have a similar question to @stitakis as the use case isn't clear to me. Would the intention for the quickstarter be for enabling [microgateways](https://apifriends.com/api-security/api-microgateway/) or for a central API gateway? For a microgateway I agree that a quickstarter would be useful but for a central gateway I was thinking that a Gradle Task that you could inject into any back-end service (Spring, Go, whatever) to virtualize it on the core gateway would be a good approach.

**Comment by gerardcl** [comment]
Hi @jonkristjan , 
yes, the idea is to provide a quickstarter, but of course a company could handle this approach in a more "centralized" way (we can think about the solo.io/Gloo option as @metmajer mentioned too).
So, indeed, both options could live together in the long run :) 

A use case/story could be (that is indeed why I proposed this quickstarter concept): a product/development team starts by implementing a simple routing gateway (e.g.: via nginx) to its different resources (identity provider, backend APIs, static files,...)...then, later on, depending on the requirements they could end up being behind a general gateway service that can for example provide a secure layer or CDN-like related services.

The concept is to going from the simpler, and if required, to the complex path, so to being agile/flexible/faster on implementing requirements on the way.
